### PR TITLE
Removing Guard Conditions

### PR DESCRIPTION
**Pull request by gbenattar**
Removing guard conditions in order to not expose failure information. See https://github.com/mortendahl/rust-paillier/issues/22.

### CODE REVIEW

**Comment by mortendahl** [comment]
I have a small concern about whether the compiler could eventually optimise the code and short-circuit some of the checks; essentially, if it would ever change the evaluation order of the terms in the or-expressions..?

Have you seen this used patterns for to get constant time elsewhere?

**Comment by omershlo** [comment]
That's a good point. I am not sure how the complier will optimize it, but It might be bad. 
What I usually do is start with fail=false and for every check in the code that fails I update flag=true. 
let mut res = true;
if check1 {res = false;}
if check2 {res = false;}
...
if checkn {res = false;}
return if flag {} else {}

File: src/proof/correct_key.rs [hunk 1]
```diff
@@ -148,19 +148,19 @@ impl ProveCorrectKey<EncryptionKey, DecryptionKey> for Paillier
...
                 .chain(&challenge.x)
                 .chain(&a)
         );
-        if challenge.e != e {
-            return Err(ProofError)
-        }
+
+        fail = (challenge.e != e) || fail;
+
+        if fail { return Err(ProofError); }
```
Comment (omershlo) [inline src/proof/correct_key.rs]: This is a branch in the code. I think we can remove this branch to the final return

**Comment by mortendahl** [comment]
actually, since we're using an iterator together with `any` this will not be constant time at all ("any() is short-circuiting; in other words, it will stop processing as soon as it finds a true") and may be further complicated by it being a parallel iterator.

one way around `any` would be to switch to `all` instead. since the first two checks are using public data only I can't see how we could leak anything from that, so it seems to me that it's only the last check on `a` that could be an issue.

**Comment by omershlo** [comment]
@mortendahl it seems that `all` is also short-circuiting: https://rust-lang-nursery.github.io/rust-cookbook/concurrency.html :  rayon::all checks in parallel whether all elements of the iterator match the predicate, and returns as soon as a non-matching element is found

**Pull request merged**